cdap-proto/src/main/java/co/cask/cdap/proto/id/EntityId.java

```diff
@@ -109,8 +109,10 @@ public final EntityType getEntity() {
     String typeString = typeAndId[0];
     EntityType type = EntityType.valueOf(typeString.toUpperCase());
     if (idClass != null && !type.getIdClass().equals(idClass)) {
+      if (!type.getIdClass().isAssignableFrom(idClass)) {
```

Comment:
better to add a comment on why this this needed. Also, to be doubly sure, perhaps add another condition in this ``if`` so it is only applicable when ``type.getIdClass.equals(ProgramId.class)``.